### PR TITLE
ADBDEV-4015 2.2.2 sync

### PR DESCRIPTION
**Pull request by deart2k**

### CODE REVIEW

**Comment by Stolb27** [comment]
@RekGRpth, as far I understand, your patch 472eb0652d7d71d8605a70b9f8f15005009ae1c9 still actual despite upstream modifications from 05da9d4 and 86ff586 because diskquota still loads shared libraries of intermediate versions during executing CREATE FUNCTION in the extension script. They solved the problem by ignoring initialization logic. But they expect that user preserve previous versions of the library.

File: src/diskquota.h [hunk 1]
```diff
@@ -217,6 +217,7 @@ struct DiskquotaDBEntry
 
 	bool inited; // this entry is inited, will set to true after the worker finish the frist run.
 	bool in_use; // this slot is in using. AKA dbid != 0
+	bool corrupted; // consider this entry as invalid to start the worker on
```
Comment (bimboterminator1) [inline src/diskquota.h]: Clang-format CI check over these changes at `diskquota.h` has failed. Should I fix it?

**Comment by RekGRpth** [comment]
@red1452, these commits https://github.com/arenadata/diskquota/commit/05da9d4dbb78373ab2c38ad576b22f637cc159ec and https://github.com/arenadata/diskquota/commit/86ff586b5f7018983e7348db06ad87c8d6f73359 seem to solve the problem https://github.com/arenadata/diskquota/pull/21#discussion_r1243163146.

**Review by RekGRpth — DISMISSED** [review]

File: src/diskquota.h (hunk 1 above)
Comment (Stolb27) [reply src/diskquota.h]: Clang-format complains to upstream changes, not your changes:

```
-	bool inited; // this entry is inited, will set to true after the worker finish the frist run.
-	bool in_use; // this slot is in using. AKA dbid != 0
+	bool inited;    // this entry is inited, will set to true after the worker finish the frist run.
+	bool in_use;    // this slot is in using. AKA dbid != 0
 	bool corrupted; // consider this entry as invalid to start the worker on
```

Comment (bimboterminator1) [reply src/diskquota.h]: Well, [here](https://github.com/arenadata/diskquota/commit/3b06e37e4882b4a3d95448df1c944ded79ff9c07#diff-2f4057fef3b75c8591966cbaf2e4f6537a89038ae371d22d2d0abd65db495831), In order to pass all the checks, I eventually edited all those lines

File: src/diskquota.c [hunk 2]
```diff
@@ -163,6 +166,40 @@ diskquota_launcher_shmem_size(void)
...
+	if (pos)
+		pos = strstr(pos, "diskquota");
+	else
+		match = false;
+	if (pos)
+		pos = strstr(pos, "update");
+	else
+		match = false;
+	pfree(query);
+	return match;
```
Comment (red1452) [inline src/diskquota.c:200]: This function is incorrect if it is necessary that `ActivePortal->sourceText` must contain word `update`. This word is not checked at this function.

Comment (RekGRpth) [reply src/diskquota.c:200]: > This word is not checked at this function.

Why?

Comment (red1452) [reply src/diskquota.c:200]: After save `pos` at line 192 there is check `pos` is not null at line 195. This is check that query contains `diskquota`. If it is true, `match` is not change to `false`. Next saving pointer of `update` to `pos` does not affect to variable `match`. And `true` will be returned even if `strstr(pos, "update");` returns `NULL`.

Comment (RekGRpth) [reply src/diskquota.c:200]: Can you explain step by step?
```
182: query = "alter extension diskquota update"
184: match = true
186: pos = "alter extension diskquota update"
188: pos = "extension diskquota update"
192: pos = "diskquota update"
196: pos = "update"
200: match = true
```

Comment (red1452) [reply src/diskquota.c:200]: If you type another word instead of `update` you will get `match = true` but expect `match = false`.

Comment (RekGRpth) [reply src/diskquota.c:200]: Exactly! If last word is not update then
```
196: pos = NULL
```
They missed another check
```c

	if (pos)
		pos = strstr(pos, "update");
	else
		match = false;
	if (!pos)
		match = false;
```

File: src/diskquota.h (hunk 1 above)
Comment (Stolb27) [reply src/diskquota.h]: I filtered out unintentionally these changes when reapply your patch in f280a884c6056ad0d89a34c80b91485cf360f3ab. Let's leave it for upstream. We shouldn't reformat each patch.

**Comment by Stolb27** [comment]
Currently, diskquota 2.2.2 is affected by greenplum-db#360. Delayed.

File: src/diskquota.c [hunk 3]
```diff
@@ -163,6 +166,40 @@ diskquota_launcher_shmem_size(void)
 	size = add_size(size, mul_size(MAX_NUM_MONITORED_DB, sizeof(struct DiskquotaDBEntry))); // hidden memory for dbArray
 	return size;
 }
+
+static bool
+is_altering_extension(void)
+{
+	if (ActivePortal == NULL) return false;
+	/* QD: When the sourceTag is T_AlterExtensionStmt, then return true */
+	if (ActivePortal->sourceTag == T_AlterExtensionStmt) return true;
```
Comment (Stolb27) [inline src/diskquota.c:175]: But we don't return false on QD otherwise. As a result, we anyway validate `sourceText` that may contain any garbage.

**Comment by Stolb27** [comment]
Closed in favor of #26

**Pull request closed without merging**